### PR TITLE
feat: Create K8S events on node completion. Closes #2274

### LINKED ISSUE #2274: Emit events to event bus
# Summary

We would like to be able to inform other systems when things happen (e.g. workflow completion).

# Motivation

Sending emails on Slack messages or Kafka.

# Proposal

We should emit Kubernetes events. These can then be consumed by the following:

* https://github.com/heptiolabs/eventrouter - ★405 - but does not seem active
* https://github.com/opsgenie/kubernetes-event-exporter ★107 new, active

## Why not Argo Events?

Argo Events...

>  is mostly useful for getting events in cluster and triggering workflows/serverless functions as the actions.
> ...
> You can set up Resource gateway that listens to k8s events and trigger a openfaas function or some lightweight workflow that can publish message to kafka.

I.e. you both need to install Argo Events, and write the plumbing - it is not "first-class" or "out-of-the-box".

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlike to be merged.
* [x] Optional. I've added My organization is added to the USERS.md.
* [ ] I've signed the CLA and required builds are green.

### CODE REVIEW

File: util/argo/audit_logger.go [hunk 1]
```diff
@@ -35,10 +35,13 @@ type ObjectRef struct {
...
 const (
-	EventReasonWorkflowRunning  = "WorkflowRunning"
-	EventReasonWorkflowSucceded = "WorkflowSucceeded"
-	EventReasonWorkflowFailed   = "WorkflowFailed"
-	EventReasonWorkflowTimedOut = "WorkflowTimedOut"
+	EventReasonWorkflowRunning       = "WorkflowRunning"
+	EventReasonWorkflowSucceeded     = "WorkflowSucceeded"
+	EventReasonWorkflowFailed        = "WorkflowFailed"
+	EventReasonWorkflowTimedOut      = "WorkflowTimedOut"
+	EventReasonWorkflowNodeSucceeded = "WorkflowNodeSucceeded"
```
Comment (alexec) [inline util/argo/audit_logger.go]: should we say "Node" or "Step"?

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -1669,12 +1669,30 @@ func (woc *wfOperationCtx) markNodePhase(nodeName string, phase wfv1.NodePhase,
 	if node.Completed() && node.FinishedAt.IsZero() {
 		node.FinishedAt = metav1.Time{Time: time.Now().UTC()}
 		woc.log.Infof("node %s finished: %s", node, node.FinishedAt)
-		woc.updated = true
```
Comment (alexec) [inline workflow/controller/operator.go]: fat finger?

File: util/argo/audit_logger.go (hunk 1 above)
Comment (simster7) [reply util/argo/audit_logger.go]: `Node`

File: workflow/controller/operator.go (hunk 2 above)
Comment (simster7) [reply workflow/controller/operator.go]: Why would you think so? `node.FinishedAt` is updated.

Comment (alexec) [reply workflow/controller/operator.go]: I actually fixed my error, the comment has lingered.

File: util/argo/audit_logger.go (hunk 1 above)
Comment (alexec) [reply util/argo/audit_logger.go]: you sure? I feel like many user would understand "step"? @jessesuen - thoughts?

Comment (simster7) [reply util/argo/audit_logger.go]: We've gone through this discussion before: https://github.com/argoproj/argo/pull/1904#discussion_r382211882

`step` can be ambiguous as it doesn't include `dags`. In my opinion, the preferable term would be "task", but that ship has sailed, unfortunately.

**Pull request merged**